### ISSUE TITLE
[Bug] React/Docgen TypeScript detection insufficient in NPM monorepo.

### ISSUE THREAD

**Issue by scott-lc** [storybookjs/builder-vite#456]
### What version of `vite` are you using?

vite@3.0.2

### System info and storybook versions

  System:
    OS: macOS 11.4
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
  Binaries:
    Node: 18.3.0 - ~/.nvm/versions/node/v18.3.0/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 8.14.0 - ~/.nvm/versions/node/v18.3.0/bin/npm
  Browsers:
    Chrome: 95.0.4638.69
    Firefox: 102.0.1
    Safari: 14.1.1

@storybook/cli: 6.5.9 

### Describe the Bug

The code [here](https://github.com/storybookjs/builder-vite/blob/a4f17ea35d16161a8dc77336213fee058346ad4b/packages/builder-vite/vite-config.ts#L175) attempts to determine if typescript is installed by reading the `package.json` file and checking for a `typescript` dependency in either `pkgJson.dependencies` or  `pkgJson.devDependencies`. (should also check `pkgJson.peerDependencies` BTW).

In an NPM monorepo (AKA [workspaces](https://docs.npmjs.com/cli/v8/using-npm/workspaces)), it is likely that any typescript dependency is defined at the root of the repository (`<root>/package.json`), rather than in the `package.json` file for a particular workspace/package (`<root>/packages/some-package/package.json`).  Therefore the determination whether or not `typescript` is being used is incorrect.

[tangentially related]

As a result, for me the `react-docgen` plugin is being referenced which, incidentally, ultimately throws an exception about a missing `@rollup/pluginutils` dependency:

```
info => Loading presets
ERR! Error: Cannot find module '@rollup/pluginutils'
ERR! Require stack:
ERR! - .../node_modules/@storybook/builder-vite/dist/plugins/react-docgen.js
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by IanVS** [comment]
Thanks.  We knew that our detection mechanism isn't perfect, but it it is designed to fail in such a way that at least does not break, but rather falls back to something that works for both typescript and non-typescript projects.  We're open to ideas (and especially PRs) for improvements.

In the meantime, it's possible to force the typescript docgen to be used by adding this to your `.storybook/main.js`:

```js
module.exports = {
  typescript: {
    reactDocgen: 'react-docgen-typescript',
  }
}
```

The `@rollup/pluginutils` error is a bug.  We use that utility in the react-docgen plugin, but it's missing from our `package.json`.

**Comment by scott-lc** [comment]
```
module.exports = {
  typescript: {
    reactDocgen: 'react-docgen-typescript',
  }
}
```

won't in my case because of the `typescriptPresent` presence check:

```
if (reactDocgen === 'react-docgen-typescript' && typescriptPresent) { ... }
```

which is fine.  I can work around it.  A far as checking whether TypeScript is in-use, perhaps it is better to just do something like `require.resolve('typescript')` and let the standard Node resolver do its thing:

```
    let typescriptPresent;

    try {
      typescriptPresent = !!require.resolve('typescript');
    } catch (e) {
      typescriptPresent = false;
    }
```

**Comment by IanVS** [comment]
We did use a check similar to that at one point, but it ended up being a problem because typescript can be brought in without actually being used.  https://github.com/storybookjs/builder-vite/pull/313.

I'll try giving this some thought though and see if we can clean it up one way or another.

**Comment by IanVS** [comment]
We've released a change to our detection logic in 0.2.1.  If you're willing to try it out, @scott-lc, I'd love to hear if it now works for you.

**Comment by scott-lc** [comment]
@IanVS 

Both manually specifying `react-docgen` as well as defaulting to `react-docgen-typescript` works for me.  Thanks.

+1
